// qbjc/dist/runtime/ansi-terminal-platform.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var ansi_escapes_1 = __importDefault(require("ansi-escapes"));
var ansi_styles_1 = __importDefault(require("ansi-styles"));
var ansi_terminal_key_code_map_1 = __importDefault(require("./ansi-terminal-key-code-map"));
var runtime_1 = require("./runtime");
/** RuntimePlatform implementing screen manipulation using ANSI escape codes.
 *
 * Ideally this functionality could be made into a mixin, but 1) it relies on this.print() etc and
 * so must be a class and 2) abstract classes don't support mixins yet:
 * https://github.com/microsoft/TypeScript/issues/35356
 */
var AnsiTerminalPlatform = /** @class */ (function (_super) {
    __extends(AnsiTerminalPlatform, _super);
    function AnsiTerminalPlatform() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AnsiTerminalPlatform.prototype.moveCursorTo = function (x, y) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.print(ansi_escapes_1.default.cursorTo(x, y))];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    AnsiTerminalPlatform.prototype.setCursorVisibility = function (isCursorVisible) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.print(isCursorVisible ? ansi_escapes_1.default.cursorShow : ansi_escapes_1.default.cursorHide)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    AnsiTerminalPlatform.prototype.clearScreen = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.print(ansi_escapes_1.default.eraseScreen)];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.moveCursorTo(0, 0)];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    AnsiTerminalPlatform.prototype.setFgColor = function (colorName) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.print(ansi_styles_1.default[colorName].open)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    AnsiTerminalPlatform.prototype.setBgColor = function (colorName) {
        return __awaiter(this, void 0, void 0, function () {
            var bgColorName;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        bgColorName = "bg".concat(colorName[0].toUpperCase()).concat(colorName.substr(1));
                        return [4 /*yield*/, this.print(ansi_styles_1.default[bgColorName].open)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    AnsiTerminalPlatform.prototype.beep = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.print(ansi_escapes_1.default.beep)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /** Translates an ANSI / VT100 key code to the DOS equivalent. */
    AnsiTerminalPlatform.prototype.translateKeyCode = function (c) {
        var _a;
        return (_a = ansi_terminal_key_code_map_1.default[c]) !== null && _a !== void 0 ? _a : c;
    };
    return AnsiTerminalPlatform;
}(runtime_1.RuntimePlatform));
exports.default = AnsiTerminalPlatform;
