// qbjc/dist/runtime/browser-platform.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrowserExecutor = exports.BrowserPlatform = void 0;
// Must polyfill global 'process' variable before importing ansi-escapes.
self.process = (_a = self.process) !== null && _a !== void 0 ? _a : require('process');
var ansi_escapes_1 = __importDefault(require("ansi-escapes"));
var ansi_terminal_platform_1 = __importDefault(require("./ansi-terminal-platform"));
var executor_1 = __importDefault(require("./executor"));
// Polyfill Buffer for iconv-lite.
self.Buffer = (_b = self.Buffer) !== null && _b !== void 0 ? _b : require('buffer').Buffer;
/** RuntimePlatform for the browser environment based on xterm.js. */
var BrowserPlatform = /** @class */ (function (_super) {
    __extends(BrowserPlatform, _super);
    function BrowserPlatform(terminal) {
        var _this = _super.call(this) || this;
        _this.terminal = terminal;
        // When running in the browser, delay is implemented as a busy loop as
        // setImmediate will block input processing. So disabling delay by default.
        _this.defaultStmtExecutionDelayUs = 0;
        return _this;
    }
    BrowserPlatform.prototype.loadCompiledModule = function (code) {
        return __awaiter(this, void 0, void 0, function () {
            var match, fn, moduleObj;
            return __generator(this, function (_a) {
                match = code.match(/^(?:#![^\n]+\n)?([\s\S]+)$/);
                if (!match) {
                    throw new Error("Empty code string");
                }
                fn = new Function('module', match[1]);
                moduleObj = { exports: {} };
                fn(moduleObj);
                return [2 /*return*/, moduleObj.exports];
            });
        });
    };
    BrowserPlatform.prototype.delay = function (delayInUs) {
        return __awaiter(this, void 0, void 0, function () {
            var t0;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(typeof window === 'undefined' ||
                            typeof window.performance === 'undefined')) return [3 /*break*/, 2];
                        return [4 /*yield*/, new Promise(function (resolve) {
                                return setTimeout(resolve, Math.round(delayInUs / 1000));
                            })];
                    case 1:
                        _a.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        t0 = performance.now();
                        while ((performance.now() - t0) * 1000 < delayInUs) { }
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    BrowserPlatform.prototype.print = function (s) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.terminal.write(s);
                return [2 /*return*/];
            });
        });
    };
    BrowserPlatform.prototype.inputLine = function () {
        return __awaiter(this, void 0, void 0, function () {
            var text, cursorIdx, initialX, cols, disposeFns;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        text = [];
                        cursorIdx = 0;
                        return [4 /*yield*/, this.getCursorPosition()];
                    case 1:
                        initialX = (_a.sent()).x;
                        return [4 /*yield*/, this.getScreenSize()];
                    case 2:
                        cols = (_a.sent()).cols;
                        disposeFns = [];
                        return [4 /*yield*/, new Promise(function (resolve) {
                                var removeOnDataListenerFn = _this.terminal.onData(function (chunk) { return __awaiter(_this, void 0, void 0, function () {
                                    return __generator(this, function (_a) {
                                        switch (chunk) {
                                            case '\r':
                                                this.print('\n');
                                                resolve();
                                                break;
                                            case '\x7f': // backspace
                                                if (text.length > 0 && cursorIdx > 0) {
                                                    text.splice(--cursorIdx, 1);
                                                    this.print(ansi_escapes_1.default.cursorBackward());
                                                    this.print(ansi_escapes_1.default.cursorSavePosition);
                                                    this.print(ansi_escapes_1.default.eraseEndLine);
                                                    this.print(text.slice(cursorIdx).join(''));
                                                    this.print(ansi_escapes_1.default.cursorRestorePosition);
                                                }
                                                break;
                                            case '\x1b[3~': // delete
                                                if (cursorIdx < text.length) {
                                                    text.splice(cursorIdx, 1);
                                                    this.print(ansi_escapes_1.default.cursorSavePosition);
                                                    this.print(ansi_escapes_1.default.eraseEndLine);
                                                    this.print(text.slice(cursorIdx).join(''));
                                                    this.print(ansi_escapes_1.default.cursorRestorePosition);
                                                }
                                                break;
                                            case '\x1b[C': // right
                                                if (cursorIdx < text.length) {
                                                    ++cursorIdx;
                                                    this.print(chunk);
                                                }
                                                break;
                                            case '\x1b[D': // left
                                                if (cursorIdx > 0) {
                                                    --cursorIdx;
                                                    this.print(chunk);
                                                }
                                                break;
                                            case '\x1b[F': // end
                                                if (cursorIdx < text.length) {
                                                    this.print(ansi_escapes_1.default.cursorForward(text.length - cursorIdx));
                                                    cursorIdx = text.length;
                                                }
                                                break;
                                            case '\x1b[H': // home
                                                if (cursorIdx > 0) {
                                                    this.print(ansi_escapes_1.default.cursorBackward(cursorIdx));
                                                    cursorIdx = 0;
                                                }
                                                break;
                                            default:
                                                // TODO: Support multi-line line editing.
                                                if (chunk.match(/^[\x20-\x7e]$/) &&
                                                    initialX + text.length < cols - 2) {
                                                    // Printable ASCII
                                                    text.splice(cursorIdx++, 0, chunk);
                                                    this.print(chunk);
                                                    this.print(ansi_escapes_1.default.cursorSavePosition);
                                                    this.print(text.slice(cursorIdx).join(''));
                                                    this.print(ansi_escapes_1.default.cursorRestorePosition);
                                                }
                                                else {
                                                    console.log(chunk
                                                        .split('')
                                                        .map(function (c) { return c.charCodeAt(0); })
                                                        .join(', '));
                                                }
                                                break;
                                        }
                                        return [2 /*return*/];
                                    });
                                }); }).dispose;
                                disposeFns.push(removeOnDataListenerFn);
                                var checkShouldStopExecutionIntervalId = setInterval(function () {
                                    if (_this.shouldStopExecution) {
                                        resolve();
                                    }
                                }, 20);
                                disposeFns.push(function () { return clearInterval(checkShouldStopExecutionIntervalId); });
                            })];
                    case 3:
                        _a.sent();
                        disposeFns.forEach(function (fn) { return fn(); });
                        return [2 /*return*/, text.join('')];
                }
            });
        });
    };
    BrowserPlatform.prototype.getChar = function () {
        return __awaiter(this, void 0, void 0, function () {
            var result, dispose;
            var _this = this;
            return __generator(this, function (_a) {
                result = null;
                dispose = this.terminal.onData(function (chunk) {
                    result = _this.translateKeyCode(chunk.toString());
                }).dispose;
                return [2 /*return*/, new Promise(function (resolve) {
                        setTimeout(function () {
                            dispose();
                            resolve(result);
                        }, 20);
                    })];
            });
        });
    };
    BrowserPlatform.prototype.getCursorPosition = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, new Promise(function (resolve, reject) {
                        var dispose = _this.terminal.onData(function (chunk) {
                            dispose();
                            var match = chunk.toString().match(/\[(\d+)\;(\d+)R/);
                            if (match) {
                                var _a = [match[1], match[2]], row = _a[0], col = _a[1];
                                var result = {
                                    // ANSI row & col numbers are 1-based.
                                    x: parseInt(col) - 1,
                                    y: parseInt(row) - 1,
                                };
                                resolve(result);
                            }
                            else {
                                reject();
                            }
                        }).dispose;
                        _this.print(ansi_escapes_1.default.cursorGetPosition);
                    })];
            });
        });
    };
    BrowserPlatform.prototype.getScreenSize = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, {
                        rows: this.terminal.rows,
                        cols: this.terminal.cols,
                    }];
            });
        });
    };
    return BrowserPlatform;
}(ansi_terminal_platform_1.default));
exports.BrowserPlatform = BrowserPlatform;
var BrowserExecutor = /** @class */ (function (_super) {
    __extends(BrowserExecutor, _super);
    function BrowserExecutor(terminal, opts) {
        if (opts === void 0) { opts = {}; }
        return _super.call(this, new BrowserPlatform(terminal), opts) || this;
    }
    return BrowserExecutor;
}(executor_1.default));
exports.BrowserExecutor = BrowserExecutor;
