// qbjc/dist/runtime/builtins.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.lookupBuiltin = exports.BUILTIN_SUBS = exports.BUILTIN_FNS = void 0;
var iconv_lite_1 = __importDefault(require("iconv-lite"));
var round_half_to_even_1 = __importDefault(require("../lib/round-half-to-even"));
var symbol_table_1 = require("../lib/symbol-table");
var types_1 = require("../lib/types");
/** Built-in functions. */
exports.BUILTIN_FNS = __spreadArray(__spreadArray([
    {
        name: 'abs',
        paramTypeSpecs: [(0, types_1.doubleSpec)()],
        returnTypeSpec: (0, types_1.doubleSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, Math.abs(n)];
                });
            });
        },
    },
    {
        name: 'asc',
        paramTypeSpecs: [(0, types_1.stringSpec)()],
        returnTypeSpec: (0, types_1.longSpec)(),
        run: function (s) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    if (s.length === 0) {
                        throw new Error('Expected non-empty string in ASC()');
                    }
                    return [2 /*return*/, iconv_lite_1.default.encode(s[0], 'cp437')[0]];
                });
            });
        },
    },
    {
        name: 'atn',
        paramTypeSpecs: [(0, types_1.doubleSpec)()],
        returnTypeSpec: (0, types_1.doubleSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, Math.atan(n)];
                });
            });
        },
    },
    {
        name: 'chr$',
        paramTypeSpecs: [(0, types_1.longSpec)()],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, iconv_lite_1.default.decode(Buffer.from([n]), 'cp437')];
                });
            });
        },
    },
    {
        name: 'cdbl',
        paramTypeSpecs: [(0, types_1.doubleSpec)()],
        returnTypeSpec: (0, types_1.doubleSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, n];
                });
            });
        },
    },
    {
        name: 'cint',
        paramTypeSpecs: [(0, types_1.doubleSpec)()],
        returnTypeSpec: (0, types_1.longSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, (0, round_half_to_even_1.default)(n)];
                });
            });
        },
    },
    {
        name: 'clng',
        paramTypeSpecs: [(0, types_1.doubleSpec)()],
        returnTypeSpec: (0, types_1.longSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, (0, round_half_to_even_1.default)(n)];
                });
            });
        },
    },
    {
        name: 'cos',
        paramTypeSpecs: [(0, types_1.doubleSpec)()],
        returnTypeSpec: (0, types_1.doubleSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, Math.cos(n)];
                });
            });
        },
    },
    {
        name: 'csng',
        paramTypeSpecs: [(0, types_1.doubleSpec)()],
        returnTypeSpec: (0, types_1.singleSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, n];
                });
            });
        },
    },
    {
        name: 'csrlin',
        paramTypeSpecs: [],
        returnTypeSpec: (0, types_1.longSpec)(),
        run: function (_a) {
            var platform = _a.platform;
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0: return [4 /*yield*/, platform.getCursorPosition()];
                        case 1: return [2 /*return*/, (_b.sent()).y + 1];
                    }
                });
            });
        },
    },
    {
        name: 'date$',
        paramTypeSpecs: [],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function () {
            return __awaiter(this, void 0, void 0, function () {
                var now;
                return __generator(this, function (_a) {
                    now = new Date();
                    return [2 /*return*/, [now.getMonth() + 1, now.getDate(), now.getFullYear()]
                            .map(function (p) { return p.toString().padStart(2, '0'); })
                            .join('-')];
                });
            });
        },
    },
    {
        name: 'exp',
        paramTypeSpecs: [(0, types_1.doubleSpec)()],
        returnTypeSpec: (0, types_1.doubleSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, Math.exp(n)];
                });
            });
        },
    },
    {
        name: 'fix',
        paramTypeSpecs: [(0, types_1.doubleSpec)()],
        returnTypeSpec: (0, types_1.longSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, Math.sign(n) * Math.floor(Math.abs(n))];
                });
            });
        },
    },
    {
        name: 'hex$',
        paramTypeSpecs: [(0, types_1.longSpec)()],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, n.toString(16).toUpperCase()];
                });
            });
        },
    },
    {
        name: 'inkey$',
        paramTypeSpecs: [],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function (_a) {
            var platform = _a.platform;
            return __awaiter(this, void 0, void 0, function () {
                var c;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0: return [4 /*yield*/, platform.getChar()];
                        case 1:
                            c = _b.sent();
                            return [2 /*return*/, c !== null && c !== void 0 ? c : ''];
                    }
                });
            });
        },
    },
    {
        name: 'input$',
        paramTypeSpecs: [(0, types_1.longSpec)()],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function (n, _a) {
            var platform = _a.platform;
            return __awaiter(this, void 0, void 0, function () {
                var result, c;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            result = [];
                            _b.label = 1;
                        case 1:
                            if (!(result.length < n && !platform.shouldStopExecution)) return [3 /*break*/, 3];
                            return [4 /*yield*/, platform.getChar()];
                        case 2:
                            c = _b.sent();
                            if (c) {
                                result.push(c);
                            }
                            return [3 /*break*/, 1];
                        case 3: return [2 /*return*/, result.join('')];
                    }
                });
            });
        },
    },
    {
        name: 'instr',
        paramTypeSpecs: [(0, types_1.stringSpec)(), (0, types_1.stringSpec)()],
        returnTypeSpec: (0, types_1.longSpec)(),
        run: function (haystack, needle) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, haystack.indexOf(needle) + 1];
                });
            });
        },
    },
    {
        name: 'instr',
        paramTypeSpecs: [(0, types_1.longSpec)(), (0, types_1.stringSpec)(), (0, types_1.stringSpec)()],
        returnTypeSpec: (0, types_1.integerSpec)(),
        run: function (start, haystack, needle) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, haystack.indexOf(needle, start - 1) + 1];
                });
            });
        },
    },
    {
        name: 'int',
        paramTypeSpecs: [(0, types_1.doubleSpec)()],
        returnTypeSpec: (0, types_1.longSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, Math.floor(n)];
                });
            });
        },
    },
    {
        name: 'lbound',
        paramTypeSpecs: [(0, types_1.arraySpec)((0, types_1.doubleSpec)(), [])],
        returnTypeSpec: (0, types_1.longSpec)(),
        run: function (array) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, array.typeSpec.dimensionSpecs[0][0]];
                });
            });
        },
    },
    {
        name: 'lbound',
        paramTypeSpecs: [(0, types_1.arraySpec)((0, types_1.doubleSpec)(), []), (0, types_1.longSpec)()],
        returnTypeSpec: (0, types_1.longSpec)(),
        run: function (array, dimensionIdx) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    if (dimensionIdx < 1 ||
                        dimensionIdx > array.typeSpec.dimensionSpecs.length) {
                        throw new Error("Invalid dimension index: ".concat(dimensionIdx));
                    }
                    return [2 /*return*/, array.typeSpec.dimensionSpecs[dimensionIdx - 1][0]];
                });
            });
        },
    },
    {
        name: 'lcase$',
        paramTypeSpecs: [(0, types_1.stringSpec)()],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function (s) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, s.toLowerCase()];
                });
            });
        },
    },
    {
        name: 'left$',
        paramTypeSpecs: [(0, types_1.stringSpec)(), (0, types_1.longSpec)()],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function (s, n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, s.substr(0, n)];
                });
            });
        },
    },
    {
        name: 'len',
        paramTypeSpecs: [(0, types_1.stringSpec)()],
        returnTypeSpec: (0, types_1.longSpec)(),
        run: function (s) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, s.length];
                });
            });
        },
    },
    {
        name: 'log',
        paramTypeSpecs: [(0, types_1.doubleSpec)()],
        returnTypeSpec: (0, types_1.doubleSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, Math.log(n)];
                });
            });
        },
    },
    {
        name: 'ltrim$',
        paramTypeSpecs: [(0, types_1.stringSpec)()],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function (s) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, s.trimStart()];
                });
            });
        },
    },
    {
        name: 'mid$',
        paramTypeSpecs: [(0, types_1.stringSpec)(), (0, types_1.longSpec)(), (0, types_1.longSpec)()],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function (s, startIdx, length) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, s.substr(startIdx - 1, length)];
                });
            });
        },
    },
    {
        name: 'oct$',
        paramTypeSpecs: [(0, types_1.longSpec)()],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, n.toString(8)];
                });
            });
        },
    },
    {
        name: 'peek',
        paramTypeSpecs: [(0, types_1.longSpec)()],
        returnTypeSpec: (0, types_1.longSpec)(),
        run: function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    // STUB
                    return [2 /*return*/, 0];
                });
            });
        },
    },
    {
        name: 'pos',
        paramTypeSpecs: [(0, types_1.longSpec)()],
        returnTypeSpec: (0, types_1.longSpec)(),
        run: function (_, _a) {
            var platform = _a.platform;
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0: return [4 /*yield*/, platform.getCursorPosition()];
                        case 1: return [2 /*return*/, (_b.sent()).x + 1];
                    }
                });
            });
        },
    },
    {
        name: 'right$',
        paramTypeSpecs: [(0, types_1.stringSpec)(), (0, types_1.longSpec)()],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function (s, n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, s.substr(Math.max(0, s.length - n))];
                });
            });
        },
    }
], overload({
    name: 'rnd',
    returnTypeSpec: (0, types_1.doubleSpec)(),
    run: function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, Math.random()];
            });
        });
    },
}, [[], [(0, types_1.longSpec)()]]), true), [
    {
        name: 'rtrim$',
        paramTypeSpecs: [(0, types_1.stringSpec)()],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function (s, _a) {
            var args = _a.args;
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_b) {
                    return [2 /*return*/, s.trimEnd()];
                });
            });
        },
    },
    {
        name: 'sgn',
        paramTypeSpecs: [(0, types_1.doubleSpec)()],
        returnTypeSpec: (0, types_1.longSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, Math.sign(n)];
                });
            });
        },
    },
    {
        name: 'sin',
        paramTypeSpecs: [(0, types_1.doubleSpec)()],
        returnTypeSpec: (0, types_1.doubleSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, Math.sin(n)];
                });
            });
        },
    },
    {
        name: 'spc',
        paramTypeSpecs: [(0, types_1.longSpec)()],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, ' '.repeat(n)];
                });
            });
        },
    },
    {
        name: 'space$',
        paramTypeSpecs: [(0, types_1.longSpec)()],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, ' '.repeat(n)];
                });
            });
        },
    },
    {
        name: 'sqr',
        paramTypeSpecs: [(0, types_1.doubleSpec)()],
        returnTypeSpec: (0, types_1.doubleSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, Math.sqrt(n)];
                });
            });
        },
    },
    {
        name: 'str$',
        paramTypeSpecs: [(0, types_1.doubleSpec)()],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, "".concat(n >= 0 ? ' ' : '').concat(n)];
                });
            });
        },
    },
    {
        name: 'string$',
        paramTypeSpecs: [(0, types_1.longSpec)(), (0, types_1.longSpec)()],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function (m, n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, iconv_lite_1.default.decode(Buffer.from([n]), 'cp437').repeat(m)];
                });
            });
        },
    },
    {
        name: 'string$',
        paramTypeSpecs: [(0, types_1.longSpec)(), (0, types_1.stringSpec)()],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function (m, s) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    if (s.length === 0) {
                        throw new Error("Empty string argument to STRING$");
                    }
                    return [2 /*return*/, s[0].repeat(m)];
                });
            });
        },
    },
    {
        name: 'tab',
        paramTypeSpecs: [(0, types_1.longSpec)()],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function (col, _a) {
            var platform = _a.platform;
            return __awaiter(this, void 0, void 0, function () {
                var _b, x, y, cols;
                return __generator(this, function (_c) {
                    switch (_c.label) {
                        case 0: return [4 /*yield*/, platform.getCursorPosition()];
                        case 1:
                            _b = _c.sent(), x = _b.x, y = _b.y;
                            return [4 /*yield*/, platform.getScreenSize()];
                        case 2:
                            cols = (_c.sent()).cols;
                            col = (Math.max(1, col) - 1) % cols;
                            return [4 /*yield*/, platform.moveCursorTo(col, x <= col ? y : y + 1)];
                        case 3:
                            _c.sent();
                            return [2 /*return*/, ''];
                    }
                });
            });
        },
    },
    {
        name: 'tan',
        paramTypeSpecs: [(0, types_1.doubleSpec)()],
        returnTypeSpec: (0, types_1.doubleSpec)(),
        run: function (n) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, Math.tan(n)];
                });
            });
        },
    },
    {
        name: 'time$',
        paramTypeSpecs: [],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function () {
            return __awaiter(this, void 0, void 0, function () {
                var now;
                return __generator(this, function (_a) {
                    now = new Date();
                    return [2 /*return*/, [now.getHours(), now.getMinutes(), now.getSeconds()]
                            .map(function (p) { return p.toString().padStart(2, '0'); })
                            .join(':')];
                });
            });
        },
    },
    {
        name: 'timer',
        paramTypeSpecs: [],
        returnTypeSpec: (0, types_1.longSpec)(),
        run: function () {
            return __awaiter(this, void 0, void 0, function () {
                var now, midnight, result;
                return __generator(this, function (_a) {
                    now = new Date();
                    midnight = new Date(now);
                    midnight.setHours(0, 0, 0, 0);
                    result = (now.getTime() - midnight.getTime()) / 1000;
                    return [2 /*return*/, result];
                });
            });
        },
    },
    {
        name: 'ubound',
        paramTypeSpecs: [(0, types_1.arraySpec)((0, types_1.doubleSpec)(), [])],
        returnTypeSpec: (0, types_1.longSpec)(),
        run: function (array) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, array.typeSpec.dimensionSpecs[0][1]];
                });
            });
        },
    },
    {
        name: 'ubound',
        paramTypeSpecs: [(0, types_1.arraySpec)((0, types_1.doubleSpec)(), []), (0, types_1.longSpec)()],
        returnTypeSpec: (0, types_1.longSpec)(),
        run: function (array, dimensionIdx) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    if (dimensionIdx < 1 ||
                        dimensionIdx > array.typeSpec.dimensionSpecs.length) {
                        throw new Error("Invalid dimension index: ".concat(dimensionIdx));
                    }
                    return [2 /*return*/, array.typeSpec.dimensionSpecs[dimensionIdx - 1][1]];
                });
            });
        },
    },
    {
        name: 'ucase$',
        paramTypeSpecs: [(0, types_1.stringSpec)()],
        returnTypeSpec: (0, types_1.stringSpec)(),
        run: function (s) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, s.toUpperCase()];
                });
            });
        },
    },
    {
        name: 'val',
        paramTypeSpecs: [(0, types_1.stringSpec)()],
        returnTypeSpec: (0, types_1.doubleSpec)(),
        run: function (s) {
            return __awaiter(this, void 0, void 0, function () {
                var v;
                return __generator(this, function (_a) {
                    v = parseFloat(s);
                    return [2 /*return*/, isNaN(v) ? 0 : v];
                });
            });
        },
    },
], false);
/** Simple statements implemented as built-in subs. */
exports.BUILTIN_SUBS = __spreadArray(__spreadArray(__spreadArray(__spreadArray(__spreadArray(__spreadArray(__spreadArray(__spreadArray(__spreadArray(__spreadArray([
    {
        name: 'beep',
        paramTypeSpecs: [],
        run: function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            return __awaiter(this, void 0, void 0, function () {
                var platform;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            platform = args.pop().platform;
                            return [4 /*yield*/, platform.beep()];
                        case 1:
                            _a.sent();
                            return [2 /*return*/];
                    }
                });
            });
        },
    }
], overload({
    name: 'cls',
    run: function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return __awaiter(this, void 0, void 0, function () {
            var platform;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        platform = args.pop().platform;
                        return [4 /*yield*/, platform.clearScreen()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    },
}, [[], [(0, types_1.longSpec)()]]), true), overload({
    name: 'color',
    run: function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return __awaiter(this, void 0, void 0, function () {
            var _a, platform, runtime, getColorName, fg, bg;
            return __generator(this, function (_b) {
                _a = args.pop(), platform = _a.platform, runtime = _a.runtime;
                getColorName = function (n) {
                    if (n < 0 || n >= runtime.currentScreenModeConfig.colorMap.length) {
                        throw new Error("Invalid color: ".concat(n));
                    }
                    return runtime.currentScreenModeConfig.colorMap[n];
                };
                fg = args[0], bg = args[1];
                if (Number.isFinite(fg)) {
                    platform.setFgColor(getColorName(fg));
                }
                if (Number.isFinite(bg)) {
                    platform.setBgColor(getColorName(bg));
                }
                return [2 /*return*/];
            });
        });
    },
}, [
    [(0, types_1.longSpec)()],
    [(0, types_1.longSpec)(), (0, types_1.longSpec)()],
    [(0, types_1.longSpec)(), (0, types_1.longSpec)(), (0, types_1.longSpec)()],
]), true), overload({
    name: '__def_seg',
    run: function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/];
            });
        });
    },
}, [[], [(0, types_1.longSpec)()]]), true), overload({
    name: 'locate',
    run: function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return __awaiter(this, void 0, void 0, function () {
            var platform, row, col, cursor, origPos, newX, newY;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        platform = args.pop().platform;
                        row = args[0], col = args[1], cursor = args[2];
                        if (!(Number.isFinite(row) && Number.isFinite(col))) return [3 /*break*/, 2];
                        return [4 /*yield*/, platform.moveCursorTo(col - 1, row - 1)];
                    case 1:
                        _a.sent();
                        return [3 /*break*/, 5];
                    case 2:
                        if (!(Number.isFinite(row) || Number.isFinite(col))) return [3 /*break*/, 5];
                        return [4 /*yield*/, platform.getCursorPosition()];
                    case 3:
                        origPos = _a.sent();
                        newX = Number.isFinite(col) ? col - 1 : origPos.x;
                        newY = Number.isFinite(row) ? row - 1 : origPos.y;
                        if (!(newX !== origPos.x || newY !== origPos.y)) return [3 /*break*/, 5];
                        return [4 /*yield*/, platform.moveCursorTo(newX, newY)];
                    case 4:
                        _a.sent();
                        _a.label = 5;
                    case 5:
                        if (!Number.isFinite(cursor)) return [3 /*break*/, 7];
                        return [4 /*yield*/, platform.setCursorVisibility(!!cursor)];
                    case 6:
                        _a.sent();
                        _a.label = 7;
                    case 7: return [2 /*return*/];
                }
            });
        });
    },
}, [
    [(0, types_1.longSpec)()],
    [(0, types_1.longSpec)(), (0, types_1.longSpec)()],
    [(0, types_1.longSpec)(), (0, types_1.longSpec)(), (0, types_1.longSpec)()],
    [(0, types_1.longSpec)(), (0, types_1.longSpec)(), (0, types_1.longSpec)(), (0, types_1.longSpec)()],
    [(0, types_1.longSpec)(), (0, types_1.longSpec)(), (0, types_1.longSpec)(), (0, types_1.longSpec)(), (0, types_1.longSpec)()],
]), true), [
    {
        name: 'play',
        paramTypeSpecs: [(0, types_1.stringSpec)()],
        run: function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/];
                });
            });
        },
    },
    {
        name: 'poke',
        paramTypeSpecs: [(0, types_1.longSpec)(), (0, types_1.longSpec)()],
        run: function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/];
                });
            });
        },
    }
], false), overload({
    name: 'randomize',
    run: function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/];
            });
        });
    },
}, [[], [(0, types_1.longSpec)()]]), true), overload({
    name: 'screen',
    run: function (mode) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        return __awaiter(this, void 0, void 0, function () {
            var runtime;
            return __generator(this, function (_a) {
                runtime = args.pop().runtime;
                runtime.setScreenMode(mode);
                return [2 /*return*/];
            });
        });
    },
}, [
    [(0, types_1.longSpec)()],
    [(0, types_1.longSpec)(), (0, types_1.longSpec)()],
    [(0, types_1.longSpec)(), (0, types_1.longSpec)(), (0, types_1.longSpec)()],
    [(0, types_1.longSpec)(), (0, types_1.longSpec)(), (0, types_1.longSpec)(), (0, types_1.longSpec)()],
]), true), overload({
    name: 'sleep',
    run: function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return __awaiter(this, void 0, void 0, function () {
            var platform, numSeconds, startTimeMs, nowMs, c;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        platform = args.pop().platform;
                        numSeconds = args.length > 0 ? args[0] : 0;
                        startTimeMs = new Date().getTime();
                        _a.label = 1;
                    case 1: return [4 /*yield*/, platform.getChar()];
                    case 2:
                        c = _a.sent();
                        nowMs = new Date().getTime();
                        _a.label = 3;
                    case 3:
                        if (c === null &&
                            (numSeconds <= 0 || nowMs - startTimeMs < numSeconds * 1000) &&
                            !platform.shouldStopExecution) return [3 /*break*/, 1];
                        _a.label = 4;
                    case 4: return [2 /*return*/];
                }
            });
        });
    },
}, [[], [(0, types_1.longSpec)()]]), true), overload({
    name: '__view_print',
    run: function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/];
            });
        });
    },
}, [[], [(0, types_1.longSpec)(), (0, types_1.longSpec)()]]), true), [
    {
        name: 'width',
        paramTypeSpecs: [(0, types_1.longSpec)(), (0, types_1.longSpec)()],
        run: function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/];
                });
            });
        },
    },
], false);
/** Helper function to generate multiple signatures that map to the same built-in proc. */
function overload(proc, paramTypeSpecsList) {
    return paramTypeSpecsList.map(function (paramTypeSpecs) {
        return (__assign(__assign({}, proc), { paramTypeSpecs: paramTypeSpecs }));
    });
}
function lookupBuiltin(procs, name, argTypeSpecs, _a) {
    var _b = _a === void 0 ? {} : _a, shouldReturnIfArgTypeMismatch = _b.shouldReturnIfArgTypeMismatch;
    var builtinFnsMatchingName = (0, symbol_table_1.lookupSymbols)(procs, name);
    var builtinFn = builtinFnsMatchingName.find(function (_a) {
        var paramTypeSpecs = _a.paramTypeSpecs;
        return paramTypeSpecs.length === argTypeSpecs.length &&
            paramTypeSpecs.every(function (paramTypeSpec, i) {
                return (0, types_1.areMatchingElementaryTypes)(paramTypeSpec, argTypeSpecs[i]) ||
                    ((0, types_1.isArray)(paramTypeSpec) && (0, types_1.isArray)(argTypeSpecs[i]));
            });
    });
    return (builtinFn !== null && builtinFn !== void 0 ? builtinFn : (shouldReturnIfArgTypeMismatch && builtinFnsMatchingName.length > 0
        ? builtinFnsMatchingName[0]
        : null));
}
exports.lookupBuiltin = lookupBuiltin;
