// qbjc/dist/runtime/executor.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExecutionError = void 0;
var data_item_1 = require("../lib/data-item");
var error_with_loc_1 = __importDefault(require("../lib/error-with-loc"));
var symbol_table_1 = require("../lib/symbol-table");
var types_1 = require("../lib/types");
var compiled_code_1 = require("./compiled-code");
var init_value_1 = __importDefault(require("./init-value"));
var runtime_1 = __importDefault(require("./runtime"));
var ExecutionError = /** @class */ (function (_super) {
    __extends(ExecutionError, _super);
    function ExecutionError(message, _a) {
        var _b = _a === void 0 ? {} : _a, module = _b.module, stmt = _b.stmt, cause = _b.cause;
        return _super.call(this, message, {
            sourceFileName: module === null || module === void 0 ? void 0 : module.sourceFileName,
            loc: (stmt === null || stmt === void 0 ? void 0 : stmt.loc) ? { line: stmt.loc[0], col: stmt.loc[1] } : undefined,
            cause: cause,
        }) || this;
    }
    return ExecutionError;
}(error_with_loc_1.default));
exports.ExecutionError = ExecutionError;
/** Exception thrown to signal program termination via END. */
var EndDirective = /** @class */ (function (_super) {
    __extends(EndDirective, _super);
    function EndDirective() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isEndDirective = true;
        return _this;
    }
    return EndDirective;
}(Error));
/** Manages the execution of a compiled program. */
var Executor = /** @class */ (function () {
    function Executor(platform, _a) {
        var _b = _a === void 0 ? {} : _a, stmtExecutionDelayUs = _b.stmtExecutionDelayUs;
        this.platform = platform;
        /** Current module being executed. */
        this.currentModule = null;
        /** Static vars, indexed by proc name. */
        this.localStaticVarsMap = {};
        /** Label index map for module-level statements. */
        this.moduleLabelIndexMap = {};
        /** Label index map for procs. */
        this.procLabelIndexMap = {};
        /** Cursor for READ. */
        this.dataCursor = {
            stmtIdx: 0,
            itemIdx: 0,
        };
        this.stmtExecutionDelayUs =
            stmtExecutionDelayUs !== null && stmtExecutionDelayUs !== void 0 ? stmtExecutionDelayUs : platform.defaultStmtExecutionDelayUs;
    }
    /** Executes a compiled module. */
    Executor.prototype.executeModule = function (moduleOrCode) {
        return __awaiter(this, void 0, void 0, function () {
            var module, _a, ctx, e_1;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!(typeof moduleOrCode === 'string')) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.platform.loadCompiledModule(moduleOrCode)];
                    case 1:
                        _a = _b.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        _a = moduleOrCode;
                        _b.label = 3;
                    case 3:
                        module = _a;
                        this.currentModule = module;
                        this.buildLabelIndexMaps();
                        this.localStaticVarsMap = {};
                        this.restore();
                        this.platform.shouldStopExecution = false;
                        return [4 /*yield*/, this.platform.setCursorVisibility(false)];
                    case 4:
                        _b.sent();
                        ctx = {
                            runtime: new runtime_1.default(this.platform),
                            executeProc: this.executeProc.bind(this),
                            read: this.read.bind(this),
                            restore: this.restore.bind(this),
                            args: {},
                            localVars: this.initVars(module.localSymbols, [
                                symbol_table_1.VarType.VAR,
                                symbol_table_1.VarType.CONST,
                            ]),
                            localStaticVars: {},
                            globalVars: this.initVars(module.globalSymbols, [
                                symbol_table_1.VarType.VAR,
                                symbol_table_1.VarType.CONST,
                            ]),
                            tempVars: {},
                        };
                        _b.label = 5;
                    case 5:
                        _b.trys.push([5, 7, 8, 10]);
                        return [4 /*yield*/, this.executeStmts(ctx, module.stmts, this.moduleLabelIndexMap)];
                    case 6:
                        _b.sent();
                        return [3 /*break*/, 10];
                    case 7:
                        e_1 = _b.sent();
                        if (e_1.isEndDirective) {
                            // Swallow error
                        }
                        else {
                            throw e_1;
                        }
                        return [3 /*break*/, 10];
                    case 8:
                        this.currentModule = null;
                        this.moduleLabelIndexMap = {};
                        this.procLabelIndexMap = {};
                        this.localStaticVarsMap = {};
                        this.restore();
                        this.platform.shouldStopExecution = false;
                        return [4 /*yield*/, this.platform.setCursorVisibility(true)];
                    case 9:
                        _b.sent();
                        return [7 /*endfinally*/];
                    case 10: return [2 /*return*/];
                }
            });
        });
    };
    /** Sets flag to stop execution. */
    Executor.prototype.stopExecution = function () {
        this.platform.shouldStopExecution = true;
    };
    Executor.prototype.executeProc = function (prevCtx, name) {
        var argPtrs = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            argPtrs[_i - 2] = arguments[_i];
        }
        return __awaiter(this, void 0, void 0, function () {
            var proc, args, i, ctx;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        proc = (0, symbol_table_1.lookupSymbol)(this.currentModule.procs, name);
                        if (!proc) {
                            throw new Error("Procedure not found: \"".concat(name, "\""));
                        }
                        // Init args.
                        if (argPtrs.length !== proc.paramSymbols.length) {
                            throw new Error('Incorrect number of arguments to ' +
                                "".concat((0, types_1.procTypeName)(proc.type), " \"").concat(proc.name, "\": ") +
                                "expected ".concat(proc.paramSymbols.length, ", got ").concat(argPtrs.length));
                        }
                        args = {};
                        for (i = 0; i < argPtrs.length; ++i) {
                            args[proc.paramSymbols[i].name] = argPtrs[i];
                        }
                        // Init static vars.
                        if (!this.localStaticVarsMap[proc.name]) {
                            this.localStaticVarsMap[proc.name] = this.initVars(proc.localSymbols, [
                                symbol_table_1.VarType.STATIC_VAR,
                            ]);
                        }
                        ctx = __assign(__assign({}, prevCtx), { args: args, localVars: this.initVars(proc.localSymbols, [symbol_table_1.VarType.VAR, symbol_table_1.VarType.CONST]), localStaticVars: this.localStaticVarsMap[proc.name], tempVars: {} });
                        if (!this.procLabelIndexMap[proc.name]) {
                            throw new Error("Label index map not found for \"".concat(proc.name, "\""));
                        }
                        return [4 /*yield*/, this.executeStmts(ctx, proc.stmts, this.procLabelIndexMap[proc.name])];
                    case 1:
                        _a.sent();
                        // Return value.
                        if ((0, types_1.isFnOrDefFn)(proc.type)) {
                            return [2 /*return*/, ctx.localVars[proc.name]];
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    Executor.prototype.initVars = function (varSymbolTable, includeVarTypes) {
        var container = {};
        for (var _i = 0, varSymbolTable_1 = varSymbolTable; _i < varSymbolTable_1.length; _i++) {
            var symbol = varSymbolTable_1[_i];
            if (!includeVarTypes.includes(symbol.varType)) {
                continue;
            }
            container[symbol.name] = (0, init_value_1.default)(symbol.typeSpec);
        }
        return container;
    };
    Executor.prototype.executeStmts = function (ctx, stmts, labelIndexMap) {
        return __awaiter(this, void 0, void 0, function () {
            var gosubStates, stmtIdx, _loop_1, this_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        gosubStates = [];
                        stmtIdx = 0;
                        _loop_1 = function () {
                            var stmt, errorArgs, directive, e_2, gotoLabel, gosubState;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0:
                                        if (this_1.platform.shouldStopExecution) {
                                            this_1.platform.shouldStopExecution = false;
                                            throw new EndDirective();
                                        }
                                        stmt = stmts[stmtIdx];
                                        errorArgs = { module: this_1.currentModule, stmt: stmt };
                                        if (!(this_1.stmtExecutionDelayUs > 0)) return [3 /*break*/, 2];
                                        return [4 /*yield*/, this_1.platform.delay(this_1.stmtExecutionDelayUs)];
                                    case 1:
                                        _b.sent();
                                        _b.label = 2;
                                    case 2:
                                        directive = void 0;
                                        if (!('run' in stmt)) return [3 /*break*/, 6];
                                        _b.label = 3;
                                    case 3:
                                        _b.trys.push([3, 5, , 6]);
                                        return [4 /*yield*/, stmt.run(ctx)];
                                    case 4:
                                        directive = _b.sent();
                                        return [3 /*break*/, 6];
                                    case 5:
                                        e_2 = _b.sent();
                                        if (e_2.isEndDirective) {
                                            throw e_2;
                                        }
                                        else {
                                            throw new ExecutionError(e_2.message, __assign(__assign({}, errorArgs), { cause: e_2 }));
                                        }
                                        return [3 /*break*/, 6];
                                    case 6:
                                        if (!directive) {
                                            ++stmtIdx;
                                            return [2 /*return*/, "continue"];
                                        }
                                        gotoLabel = function (label) {
                                            stmtIdx = labelIndexMap[label];
                                            if (stmtIdx === undefined) {
                                                throw new ExecutionError("Label not found: \"".concat(label, "\""), errorArgs);
                                            }
                                        };
                                        switch (directive.type) {
                                            case compiled_code_1.ExecutionDirectiveType.GOSUB:
                                                gosubStates.push({
                                                    nextStmtIdx: stmtIdx + 1,
                                                });
                                                gotoLabel(directive.destLabel);
                                                break;
                                            case compiled_code_1.ExecutionDirectiveType.GOTO:
                                                gotoLabel(directive.destLabel);
                                                break;
                                            case compiled_code_1.ExecutionDirectiveType.RETURN:
                                                if (gosubStates.length === 0) {
                                                    throw new ExecutionError("RETURN without prior GOSUB", errorArgs);
                                                }
                                                gosubState = gosubStates.pop();
                                                if (directive.destLabel) {
                                                    gotoLabel(directive.destLabel);
                                                }
                                                else {
                                                    stmtIdx = gosubState.nextStmtIdx;
                                                }
                                                break;
                                            case compiled_code_1.ExecutionDirectiveType.END:
                                                throw new EndDirective();
                                            case compiled_code_1.ExecutionDirectiveType.EXIT_PROC:
                                                stmtIdx = stmts.length;
                                                break;
                                            default:
                                                throw new ExecutionError("Unrecognized execution directive: ".concat(JSON.stringify(directive)), errorArgs);
                                        }
                                        return [2 /*return*/];
                                }
                            });
                        };
                        this_1 = this;
                        _a.label = 1;
                    case 1:
                        if (!(stmtIdx < stmts.length)) return [3 /*break*/, 3];
                        return [5 /*yield**/, _loop_1()];
                    case 2:
                        _a.sent();
                        return [3 /*break*/, 1];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    Executor.prototype.read = function () {
        var resultTypes = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            resultTypes[_i] = arguments[_i];
        }
        return __awaiter(this, void 0, void 0, function () {
            var results, i, dataItem;
            return __generator(this, function (_a) {
                results = [];
                for (i = 0; i < resultTypes.length; ++i) {
                    dataItem = this.getNextDataItem();
                    if (!dataItem) {
                        throw new Error("Data exhausted at item ".concat(i + 1));
                    }
                    try {
                        results.push((0, data_item_1.getDataItem)(dataItem, resultTypes[i]));
                    }
                    catch (e) {
                        e.message = "Error reading item ".concat(i + 1, ": ").concat(e.message);
                        throw e;
                    }
                }
                return [2 /*return*/, results];
            });
        });
    };
    Executor.prototype.getNextDataItem = function () {
        var stmts = this.currentModule.stmts;
        for (;;) {
            if (this.dataCursor.stmtIdx >= stmts.length) {
                return null;
            }
            var stmt = stmts[this.dataCursor.stmtIdx];
            if (!('data' in stmt)) {
                ++this.dataCursor.stmtIdx;
                continue;
            }
            if (this.dataCursor.itemIdx >= stmt.data.length) {
                ++this.dataCursor.stmtIdx;
                this.dataCursor.itemIdx = 0;
                continue;
            }
            return stmt.data[this.dataCursor.itemIdx++];
        }
    };
    Executor.prototype.restore = function (destLabel) {
        return __awaiter(this, void 0, void 0, function () {
            var stmtIdx;
            var _a, _b;
            return __generator(this, function (_c) {
                if (destLabel) {
                    stmtIdx = this.moduleLabelIndexMap[destLabel];
                    if (stmtIdx === undefined) {
                        throw new Error("Label not found: \"".concat(destLabel, "\""));
                    }
                    _a = [stmtIdx, 0], this.dataCursor.stmtIdx = _a[0], this.dataCursor.itemIdx = _a[1];
                }
                else {
                    _b = [0, 0], this.dataCursor.stmtIdx = _b[0], this.dataCursor.itemIdx = _b[1];
                }
                return [2 /*return*/];
            });
        });
    };
    Executor.prototype.buildLabelIndexMaps = function () {
        this.moduleLabelIndexMap = this.buildLabelIndexMap(this.currentModule.stmts);
        for (var _i = 0, _a = this.currentModule.procs; _i < _a.length; _i++) {
            var _b = _a[_i], name_1 = _b.name, stmts = _b.stmts;
            this.procLabelIndexMap[name_1] = this.buildLabelIndexMap(stmts);
        }
    };
    Executor.prototype.buildLabelIndexMap = function (stmts) {
        var _this = this;
        var labelIndexMap = {};
        stmts.forEach(function (stmt, idx) {
            if ('label' in stmt) {
                if (stmt.label in labelIndexMap) {
                    throw new ExecutionError("Duplicate label \"".concat(stmt.label, "\""), {
                        module: _this.currentModule,
                        stmt: stmt,
                    });
                }
                labelIndexMap[stmt.label] = idx;
            }
        });
        return labelIndexMap;
    };
    return Executor;
}());
exports.default = Executor;
