// qbjc/dist/runtime/runtime.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SCREEN_MODE_CONFIGS = exports.RuntimePlatform = void 0;
var moo_1 = __importDefault(require("moo"));
var types_1 = require("../lib/types");
var builtins_1 = require("./builtins");
var compiled_code_1 = require("./compiled-code");
var round_half_to_even_1 = __importDefault(require("../lib/round-half-to-even"));
/** Interface for platform-specific runtime functionality. */
var RuntimePlatform = /** @class */ (function () {
    function RuntimePlatform() {
        /** Flag indicating we should stop executing.
         *
         * Setting this flag will cause blocking methods like `inputLine()` to exit early.
         */
        this.shouldStopExecution = false;
    }
    return RuntimePlatform;
}());
exports.RuntimePlatform = RuntimePlatform;
exports.SCREEN_MODE_CONFIGS = [
    {
        mode: 0,
        textModeDimensions: [{ rows: 80, cols: 25 }],
        colorMap: [
            'black',
            'blue',
            'green',
            'cyan',
            'red',
            'magenta',
            'yellow',
            'white',
            'gray',
            'blueBright',
            'greenBright',
            'cyanBright',
            'redBright',
            'magentaBright',
            'yellowBright',
            'whiteBright',
        ],
    },
];
/** Runtime support library.
 *
 * This class provides the interface invoked by compiled code. It is platform-agnostic, and wraps
 * the low-level, platform-dependent functionality in the injected RuntimePlatform.
 */
var Runtime = /** @class */ (function () {
    function Runtime(platform) {
        this.platform = platform;
        this.roundHalfToEven = round_half_to_even_1.default;
        this.printFormatStringLexer = moo_1.default.compile({
            NUMBER: /(?:[#,])*\.#+|[#,]+/,
            STRING: '&',
            DOLLAR_SIGNS: /\$+/,
            LITERAL: {
                match: /[^.#&\$]+/,
                lineBreaks: true,
            },
        });
        /** Lexer for input(). */
        this.inputLexer = moo_1.default.compile({
            WHITESPACE: { match: /\s+/, lineBreaks: true },
            QUOTED_STRING: {
                match: /"[^"]*"/,
                value: function (text) { return text.substr(1, text.length - 2); },
                type: function () { return 'STRING'; },
            },
            COMMA: ',',
            STRING: /[^,\s]+/,
        });
        this.currentScreenMode = 0;
    }
    Runtime.prototype.executeBuiltinProc = function (name, argTypeSpecs) {
        var args = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            args[_i - 2] = arguments[_i];
        }
        return __awaiter(this, void 0, void 0, function () {
            var builtinProc, runContext;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        builtinProc = (0, builtins_1.lookupBuiltin)(builtins_1.BUILTIN_FNS, name, argTypeSpecs) ||
                            (0, builtins_1.lookupBuiltin)(builtins_1.BUILTIN_SUBS, name, argTypeSpecs);
                        if (!builtinProc) {
                            throw new Error("No matching built-in procedure found: \"".concat(name, "\""));
                        }
                        if (argTypeSpecs.length !== args.length) {
                            throw new Error("Argument types and values do not match: " +
                                "expected ".concat(argTypeSpecs.length, " arguments, got ").concat(args.length));
                        }
                        runContext = {
                            runtime: this,
                            platform: this.platform,
                            argTypeSpecs: argTypeSpecs,
                            args: args,
                        };
                        return [4 /*yield*/, builtinProc.run.apply(builtinProc, __spreadArray(__spreadArray([], args.map(function (ptr) { return ptr[0][ptr[1]]; }), false), [runContext], false))];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Runtime.prototype.print = function (formatString) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        return __awaiter(this, void 0, void 0, function () {
            var line;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        line = formatString
                            ? this.printWithFormatString(formatString, args)
                            : this.printWithDefaultFormat(args);
                        if (args.length === 0 ||
                            args[args.length - 1].type === compiled_code_1.PrintArgType.VALUE) {
                            line += '\n';
                        }
                        return [4 /*yield*/, this.platform.print(line)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /** Implements PRINT USING format strings.
     *
     * This is a VERY incomplete implementation. Supported:
     *
     *   - Formatting numbers with ##.##
     *   - Formatting string with &
     */
    Runtime.prototype.printWithFormatString = function (formatString, args) {
        var pieces = [];
        var argIdx = 0;
        this.printFormatStringLexer.reset(formatString);
        for (;;) {
            while (argIdx < args.length && args[argIdx].type !== compiled_code_1.PrintArgType.VALUE) {
                ++argIdx;
            }
            var nextArg = argIdx < args.length ? args[argIdx] : null;
            var token = this.printFormatStringLexer.next();
            if (!token) {
                break;
            }
            switch (token.type) {
                case 'NUMBER':
                    if (!nextArg || typeof nextArg.value !== 'number') {
                        throw new Error('Formatting error: expected number ' +
                            "as argument #".concat(argIdx + 1, ", got ").concat(typeof (nextArg === null || nextArg === void 0 ? void 0 : nextArg.value)));
                    }
                    var _a = token.value.split('.'), integralFormat = _a[0], fractionFormat = _a[1];
                    // Not supporting commas yet.
                    var integralFormatLength = integralFormat.replace(/,/, '').length;
                    if (fractionFormat) {
                        var formattedValue = "".concat(Math.floor(nextArg.value)).padStart(integralFormatLength) +
                            '.' +
                            nextArg.value.toFixed(fractionFormat.length).split('.')[1];
                        pieces.push(formattedValue);
                    }
                    else {
                        var formattedValue = nextArg.value.toFixed(0).padStart(integralFormatLength) +
                            (token.value.indexOf('.') >= 0 ? '.' : '');
                        pieces.push(formattedValue);
                    }
                    ++argIdx;
                    break;
                case 'STRING':
                    if (!nextArg || typeof nextArg.value !== 'string') {
                        throw new Error("Formatting error: expected string as argument #".concat(argIdx + 1));
                    }
                    pieces.push(nextArg.value);
                    ++argIdx;
                    break;
                case 'DOLLAR_SIGNS': {
                    var value = void 0;
                    if (token.value.length <= 1) {
                        value = token.value;
                    }
                    else if (token.value.length === 2) {
                        value = ' $';
                    }
                    else {
                        value = '%$';
                    }
                    pieces.push(value);
                    break;
                }
                case 'LITERAL':
                    pieces.push(token.value);
                    break;
                default:
                    throw new Error("Unexpected token type: ".concat(JSON.stringify(token)));
            }
        }
        return pieces.join('');
    };
    Runtime.prototype.printWithDefaultFormat = function (args) {
        var PRINT_ZONE_LENGTH = 14;
        var line = '';
        for (var _i = 0, args_1 = args; _i < args_1.length; _i++) {
            var arg = args_1[_i];
            switch (arg.type) {
                case compiled_code_1.PrintArgType.SEMICOLON:
                    break;
                case compiled_code_1.PrintArgType.COMMA:
                    var numPaddingChars = PRINT_ZONE_LENGTH - (line.length % PRINT_ZONE_LENGTH);
                    line += ' '.repeat(numPaddingChars);
                    break;
                case compiled_code_1.PrintArgType.VALUE:
                    if (typeof arg.value === 'string') {
                        line += "".concat(arg.value);
                        break;
                    }
                    else if (typeof arg.value === 'number') {
                        line += "".concat(arg.value >= 0 ? ' ' : '').concat(arg.value, " ");
                        break;
                    }
                    else {
                        // Fall through
                    }
                default:
                    throw new Error("Unknown print arg: '".concat(JSON.stringify(arg), "'"));
            }
        }
        return line;
    };
    Runtime.prototype.inputLine = function (prompt) {
        return __awaiter(this, void 0, void 0, function () {
            var line;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!prompt) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.platform.print(prompt)];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [4 /*yield*/, this.platform.setCursorVisibility(true)];
                    case 3:
                        _a.sent();
                        return [4 /*yield*/, this.platform.inputLine()];
                    case 4:
                        line = _a.sent();
                        return [4 /*yield*/, this.platform.setCursorVisibility(false)];
                    case 5:
                        _a.sent();
                        return [2 /*return*/, line];
                }
            });
        });
    };
    Runtime.prototype.input = function (prompt) {
        var resultTypes = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            resultTypes[_i - 1] = arguments[_i];
        }
        return __awaiter(this, void 0, void 0, function () {
            var results, line, tokens, tokenIdx, errorMessage, resultIdx, resultType, errorMessagePrefix, tokenValue, numericValue;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        results = [];
                        return [4 /*yield*/, this.platform.print(prompt)];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.inputLine('')];
                    case 2:
                        line = _a.sent();
                        if (this.platform.shouldStopExecution) {
                            return [2 /*return*/, results];
                        }
                        tokens = this.lexInput(line);
                        tokenIdx = 0;
                        errorMessage = null;
                        for (resultIdx = 0; resultIdx < resultTypes.length; ++resultIdx) {
                            resultType = resultTypes[resultIdx];
                            errorMessagePrefix = "Error parsing value ".concat(resultIdx + 1, ": ");
                            // Consume comma token for result 1+.
                            if (resultIdx > 0) {
                                if (tokenIdx < tokens.length && tokens[tokenIdx].type === 'COMMA') {
                                    ++tokenIdx;
                                }
                                else {
                                    errorMessage = "".concat(errorMessagePrefix, "Comma expected");
                                    break;
                                }
                            }
                            // Consume value.
                            if (tokenIdx < tokens.length && tokens[tokenIdx].type === 'STRING') {
                                tokenValue = tokens[tokenIdx].value;
                                if ((0, types_1.isString)(resultType)) {
                                    results.push(tokenValue);
                                }
                                else if ((0, types_1.isNumeric)(resultType)) {
                                    numericValue = parseFloat(tokenValue);
                                    if (isNaN(numericValue)) {
                                        errorMessage = "".concat(errorMessagePrefix, "Invalid numeric value \"").concat(tokenValue, "\"");
                                        break;
                                    }
                                    else {
                                        results.push(numericValue);
                                    }
                                }
                                else {
                                    throw new Error("Unexpected result type ".concat(resultType.type));
                                }
                                ++tokenIdx;
                            }
                            else {
                                errorMessage = "".concat(errorMessagePrefix, "No value provided");
                                break;
                            }
                        }
                        if (!(results.length === resultTypes.length)) return [3 /*break*/, 3];
                        return [2 /*return*/, results];
                    case 3:
                        if (!(results.length < resultTypes.length)) return [3 /*break*/, 5];
                        return [4 /*yield*/, this.platform.print("\n".concat(errorMessage ? "".concat(errorMessage, "\n") : '', "Redo from start\n"))];
                    case 4:
                        _a.sent();
                        return [3 /*break*/, 6];
                    case 5: 
                    // This would indicate a logic error above - we should stop consuming from input once the
                    // expected number of results has been reached.
                    throw new Error("Too many results: expected ".concat(resultTypes.length, ", got ").concat(results.length));
                    case 6: return [3 /*break*/, 0];
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    Runtime.prototype.lexInput = function (line) {
        this.inputLexer.reset(line.trim());
        var tokens = [];
        for (;;) {
            var token = this.inputLexer.next();
            if (!token) {
                break;
            }
            if (token.type !== 'WHITESPACE') {
                tokens.push(token);
            }
        }
        return tokens;
    };
    Runtime.prototype.setScreenMode = function (n) {
        var screenMode = exports.SCREEN_MODE_CONFIGS.find(function (_a) {
            var mode = _a.mode;
            return mode === n;
        });
        if (!screenMode) {
            throw new Error("Unsupported screen mode: ".concat(n));
        }
        this.currentScreenMode = n;
    };
    Object.defineProperty(Runtime.prototype, "currentScreenModeConfig", {
        get: function () {
            var _this = this;
            return exports.SCREEN_MODE_CONFIGS.find(function (_a) {
                var mode = _a.mode;
                return mode === _this.currentScreenMode;
            });
        },
        enumerable: false,
        configurable: true
    });
    return Runtime;
}());
exports.default = Runtime;
