// qbjc/dist/compile.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var terser_1 = require("terser");
var codegen_1 = __importDefault(require("./codegen/codegen"));
var parser_1 = __importDefault(require("./parser/parser"));
var semantic_analysis_1 = __importDefault(require("./semantic-analysis/semantic-analysis"));
var DEFAULT_SOURCE_FILE_NAME = 'source.bas';
/** Compiles a QBasic program.
 *
 * This is the main entrypoint to qbjc's compiler.
 */
function compile(_a) {
    var source = _a.source, _b = _a.sourceFileName, sourceFileName = _b === void 0 ? DEFAULT_SOURCE_FILE_NAME : _b, enableBundling = _a.enableBundling, enableMinify = _a.enableMinify;
    return __awaiter(this, void 0, void 0, function () {
        var astModule, _c, code, sourceMap, sourceMapContent, _d, minifiedCode, minifiedSourceMap;
        return __generator(this, function (_e) {
            switch (_e.label) {
                case 0:
                    astModule = (0, parser_1.default)(source, { sourceFileName: sourceFileName });
                    if (!astModule) {
                        throw new Error("Invalid parse tree");
                    }
                    // 2. Semantic analysis.
                    (0, semantic_analysis_1.default)(astModule, { sourceFileName: sourceFileName });
                    _c = (0, codegen_1.default)(astModule, {
                        sourceFileName: sourceFileName,
                        enableBundling: enableBundling,
                    }), code = _c.code, sourceMap = _c.map;
                    sourceMapContent = sourceMap.toString();
                    if (!enableMinify) return [3 /*break*/, 2];
                    return [4 /*yield*/, (0, terser_1.minify)(code, {
                            sourceMap: {
                                filename: sourceFileName,
                                content: sourceMapContent,
                            },
                        })];
                case 1:
                    _d = _e.sent(), minifiedCode = _d.code, minifiedSourceMap = _d.map;
                    if (minifiedCode && minifiedSourceMap) {
                        code = minifiedCode;
                        sourceMapContent = minifiedSourceMap;
                    }
                    _e.label = 2;
                case 2: return [2 /*return*/, {
                        source: source,
                        sourceFileName: sourceFileName,
                        code: code,
                        map: sourceMapContent,
                        astModule: astModule,
                    }];
            }
        });
    });
}
exports.default = compile;
