// qbjc/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QbUdt = exports.QbArray = exports.RuntimePlatform = exports.Executor = exports.ExecutionError = exports.compile = void 0;
var compile_1 = require("./compile");
Object.defineProperty(exports, "compile", { enumerable: true, get: function () { return __importDefault(compile_1).default; } });
__exportStar(require("./lib/error-with-loc"), exports);
__exportStar(require("./lib/types"), exports);
var executor_1 = require("./runtime/executor");
Object.defineProperty(exports, "ExecutionError", { enumerable: true, get: function () { return executor_1.ExecutionError; } });
Object.defineProperty(exports, "Executor", { enumerable: true, get: function () { return __importDefault(executor_1).default; } });
var runtime_1 = require("./runtime/runtime");
Object.defineProperty(exports, "RuntimePlatform", { enumerable: true, get: function () { return runtime_1.RuntimePlatform; } });
var qb_array_1 = require("./runtime/qb-array");
Object.defineProperty(exports, "QbArray", { enumerable: true, get: function () { return __importDefault(qb_array_1).default; } });
var qb_udt_1 = require("./runtime/qb-udt");
Object.defineProperty(exports, "QbUdt", { enumerable: true, get: function () { return __importDefault(qb_udt_1).default; } });
__exportStar(require("./runtime/compiled-code"), exports);
