// qbjc/dist/qbjc.js
#!/usr/bin/env node
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compileFile = void 0;
var commander_1 = require("commander");
var fs_extra_1 = __importDefault(require("fs-extra"));
var path_1 = __importDefault(require("path"));
var compile_1 = __importDefault(require("./compile"));
var node_platform_1 = require("./runtime/node-platform");
// Not using resolveJsonModule because it causes the output to be generated relative to the root
// directory instead of src/.
var packageJson = require('../package.json');
/** Compiles a QBasic source file and write out the compiled program and source map. */
function compileFile(_a) {
    var sourceFilePath = _a.sourceFilePath, outputFilePathArg = _a.outputFilePath, enableSourceMap = _a.enableSourceMap, enableBundling = _a.enableBundling, enableMinify = _a.enableMinify;
    return __awaiter(this, void 0, void 0, function () {
        var source, sourceFileName, compileResult, code, sourceMapContent, outputFilePath, sourceMapFileName;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0: return [4 /*yield*/, fs_extra_1.default.readFile(sourceFilePath, 'utf-8')];
                case 1:
                    source = _b.sent();
                    sourceFileName = path_1.default.basename(sourceFilePath);
                    return [4 /*yield*/, (0, compile_1.default)({
                            source: source,
                            sourceFileName: sourceFileName,
                            enableBundling: enableBundling,
                            enableMinify: enableMinify,
                        })];
                case 2:
                    compileResult = _b.sent();
                    code = compileResult.code, sourceMapContent = compileResult.map;
                    outputFilePath = outputFilePathArg || "".concat(sourceFilePath, ".js");
                    if (!enableSourceMap) return [3 /*break*/, 4];
                    sourceMapFileName = "".concat(path_1.default.basename(outputFilePath), ".map");
                    return [4 /*yield*/, fs_extra_1.default.writeFile(path_1.default.join(path_1.default.dirname(outputFilePath), sourceMapFileName), sourceMapContent)];
                case 3:
                    _b.sent();
                    code += "\n//# sourceMappingURL=".concat(sourceMapFileName, "\n");
                    _b.label = 4;
                case 4: return [4 /*yield*/, fs_extra_1.default.writeFile(outputFilePath, code)];
                case 5:
                    _b.sent();
                    if (!enableBundling) return [3 /*break*/, 7];
                    return [4 /*yield*/, fs_extra_1.default.chmod(outputFilePath, '755')];
                case 6:
                    _b.sent();
                    _b.label = 7;
                case 7: return [2 /*return*/, __assign(__assign({}, compileResult), { source: source, code: code, map: sourceMapContent, outputFilePath: outputFilePath })];
            }
        });
    });
}
exports.compileFile = compileFile;
// Entrypoint for the "qbjc" CLI tool.
if (require.main === module) {
    (function () { return __awaiter(void 0, void 0, void 0, function () {
        var sourceFilePath, opts, compileFileResult, executor;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    commander_1.program
                        .name(packageJson.name)
                        .version(packageJson.version)
                        .arguments('<file.bas>')
                        .option('-o, --output <file>', 'output file path')
                        .option('-r, --run', 'run the compiled program after compilation')
                        .option('--minify', 'minify the compiled program')
                        .option('--source-map', 'enable source map generation')
                        .option('--no-bundle', 'disable bundling with runtime code')
                        .option('--debug-ast', 'enable generation of AST file for debugging compilation')
                        .option('--debug-trace', "enable stack trace for debugging compilation")
                        .parse();
                    if (commander_1.program.args.length === 0) {
                        console.error('Error: No input files specified.');
                        process.exit(1);
                    }
                    else if (commander_1.program.args.length > 1) {
                        console.error('Error: Please provide a single input file.');
                        process.exit(1);
                    }
                    sourceFilePath = commander_1.program.args[0];
                    opts = commander_1.program.opts();
                    return [4 /*yield*/, compileFile({
                            sourceFilePath: sourceFilePath,
                            outputFilePath: opts.output,
                            enableSourceMap: opts.sourceMap,
                            enableBundling: opts.bundle,
                            enableMinify: opts.minify,
                        })];
                case 1:
                    compileFileResult = _a.sent();
                    if (!opts.debugAst) return [3 /*break*/, 3];
                    return [4 /*yield*/, fs_extra_1.default.writeJson("".concat(sourceFilePath, ".ast.json"), compileFileResult.astModule, {
                            spaces: 4,
                        })];
                case 2:
                    _a.sent();
                    _a.label = 3;
                case 3:
                    if (!opts.run) return [3 /*break*/, 5];
                    executor = new node_platform_1.NodeExecutor();
                    return [4 /*yield*/, executor.executeModule(compileFileResult.code)];
                case 4:
                    _a.sent();
                    _a.label = 5;
                case 5: return [2 /*return*/];
            }
        });
    }); })().catch(function (e) {
        if ('message' in e) {
            if (commander_1.program.opts().debugTrace) {
                console.trace(e);
            }
            else {
                console.error(e.message);
            }
        }
        process.exit(1);
    });
}
